cub/test/catch2_test_device_radix_sort_keys.cu

```diff
@@ -0,0 +1,471 @@
+  , half_t
+#endif
+#ifdef TEST_BF_T
+  , bfloat16_t
+#endif
+  >;
+// clang-format on
+#elif TEST_KEY_BYTES == 32
+using key_types = c2h::type_list<cuda::std::uint32_t, cuda::std::int32_t, float>;
+#elif TEST_KEY_BYTES == 64
+using key_types = c2h::type_list<cuda::std::uint64_t, cuda::std::int64_t, double>;
+#endif
+
+// Used for testing begin/end bit slicing
+template <typename T>
+struct supports_bit_window
+{
+  static constexpr bool value = cuda::std::is_unsigned<T>::value && !cuda::std::is_same<bool, T>::value;
+};
+using bit_window_key_types = c2h::copy_if<key_types, c2h::trait<supports_bit_window>>;
+
+// Used for testing floating-point specific behavior
+using fp_key_types = c2h::copy_if<key_types, c2h::trait<cuda::std::is_floating_point>>;
+
+// Used for tests that just need a single type for testing:
+using single_key_type = c2h::type_list<c2h::get<0, key_types>>;
+
+// Index types used for NumItemsT testing
+using num_items_types =
+  c2h::type_list<cuda::std::uint32_t, cuda::std::int32_t, cuda::std::uint64_t, cuda::std::int64_t>;
+
+// Helpers to assist with specifying default args to DeviceRadixSort API:
+template <typename T>
+constexpr int begin_bit()
+{
+  return 0;
+}
+
+template <typename T>
+constexpr int end_bit()
+{
+  return static_cast<int>(sizeof(T) * CHAR_BIT);
+}
+
+CUB_TEST("DeviceRadixSort::SortKeys: basic testing", "[keys][radix][sort][device]", key_types)
+{
+  using key_t = c2h::get<0, TestType>;
+
+  // Test different magnitudes:
+  const int power_of_two = GENERATE(0, 5, 10, 20);
+  // Test slightly less than a power of two, exactly a power of two, and slightly more:
+  const int offset            = GENERATE(-1, 0, 3);
+  const std::size_t num_items = (1 << power_of_two) + offset;
+
+  thrust::device_vector<key_t> in_keys(num_items);
+  thrust::device_vector<key_t> out_keys(num_items);
+
+  const int num_key_seeds = 3;
+  c2h::gen(CUB_SEED(num_key_seeds), in_keys);
+
+  const bool is_descending = GENERATE(false, true);
+
+  auto ref_keys = radix_sort_reference(in_keys, is_descending);
+
+  if (is_descending)
+  {
+    sort_keys_descending(
+      thrust::raw_pointer_cast(in_keys.data()),
+      thrust::raw_pointer_cast(out_keys.data()),
+      num_items,
+      begin_bit<key_t>(),
+      end_bit<key_t>());
+  }
+  else
+  {
+    sort_keys(thrust::raw_pointer_cast(in_keys.data()),
+              thrust::raw_pointer_cast(out_keys.data()),
+              num_items,
+              begin_bit<key_t>(),
+              end_bit<key_t>());
+  }
+
+  REQUIRE(ref_keys == out_keys);
+}
+
+CUB_TEST("DeviceRadixSort::SortKeys: bit windows", "[keys][radix][sort][device]", bit_window_key_types)
+{
+  using key_t = c2h::get<0, TestType>;
+
+  const std::size_t num_items = (1 << 18) + 7;
+
+  constexpr int num_bits = sizeof(key_t) * CHAR_BIT;
+  const int begin_bit    = GENERATE_COPY(0, 1, num_bits / 3, 3 * num_bits / 4, num_bits);
+  const int end_bit      = GENERATE_COPY(0, 1, num_bits / 3, 3 * num_bits / 4, num_bits);
```

Comment:
suggestion: same motivation as with `num_items`:
```suggestion
  const int begin_bit = GENERATE_COPY(take(2, random(0, num_bits)));
  const int end_bit   = GENERATE_COPY(take(2, random(begin_bit, num_bits)));
```

Reply:
This is generating cases where `end_bit < begin_bit`, likely do to Catch2 generator lifetime issues -- it doesn't seem to be updating the generator expressions as expected, even with `GENERATE_COPY`. I don't think Catch2 expects the values of the captured variables to be changed dynamically like this.

I'll leave this as is for now.

Reply:
@allisonvacanti I was not able to reproduce this:

```cpp
#define CATCH_CONFIG_MAIN

#include <catch2/catch.hpp>

TEST_CASE("Test", "[test]") {
  const int num_bits = 64;
  const int dummy = GENERATE_COPY(range(0, num_bits));
  const int begin_bit = GENERATE_COPY(take(300, random(0, num_bits)));
  const int end_bit = GENERATE_COPY(take(300, random(begin_bit, num_bits)));

  if (end_bit < begin_bit) {
    INFO("begin_bit = " << begin_bit);
    INFO("end_bit = " << end_bit);
    FAIL("end_bit < begin_bit");
  }
}
```

Do you have a reproducer?

Reply:
IIRC I was using something like this:

```
  const int begin_bit = GENERATE_COPY(take(2, random(0, num_bits)));
  const int end_bit = GENERATE_COPY(begin_bit, take(2, random(begin_bit, num_bits)));
```

to make sure that we'd always cover the begin_bit == end_bit edgecase.